### ISSUE TITLE
✍️ Contribution period: Sarima Chiorlu

### ISSUE THREAD

**Issue by Richiio** [ersilia-os/ersilia#828]
### Week 1 - Get to know the community

- [x] Join the communication channels
- [x] Open a GitHub issue (this one!)
- [x] Install the Ersilia Model Hub and test the simplest model
- [x] Write a motivation statement to work at Ersilia
- [x] Submit your first contribution to the Outreachy site

### Week 2 - Install and run an ML model

- [x] Select a model from the suggested list
- [x] Install the model in your system
- [x] Run predictions for the EML
- [x] Compare results with the Ersilia Model Hub implementation!
- [x] Install and run Docker!

### Week 3 - Propose new models

- [x] Suggest a new model and document it (1)
- [x] Suggest a new model and document it (2)
- [x] Suggest a new model and document it (3)

### Week 4 - Prepare your final application

- [x] Submit the final application in the Outreachy website

**Comment by Richiio** [comment]
**Installing Ersilia Model Hub**
I was able to successfully install the ersilia model hub without any issues. To test if it was installed successfully, I ran the following commands and here are the output

The output for `ersilia --help`
Usage: ersilia [OPTIONS] COMMAND [ARGS]...

  🦠 Welcome to Ersilia! 💊

Options:
  --version      Show the version and exit.
  -v, --verbose  Show logging on terminal when running commands.
  -s, --silent   Do not echo any progress message.
  --help         Show this message and exit.

Commands:
  api      Run API on a served model
  auth     Log in to ersilia to enter contributor mode.
  card     Get model info card
  catalog  List a catalog of models
  clear    Clear ersilia
  close    Close model
  current  Get identifier of current model
  delete   Delete model from local computer
  example  Generate input examples for the model of interest
  fetch    Fetch model from Ersilia Model Hub
  info     Get model information
  run      Run a served model
  sample   Sample inputs and model identifiers
  serve    Serve model
  test     Test a model

**Comment by DhanshreeA** [comment]
Hi @Richiio thanks for the updates. Please proceed to test the simplest model (eos3b5e) as mentioned in the instructions and report your progress/any issues you run into here.

**Comment by Richiio** [comment]
Alright @DhanshreeA

**Comment by Richiio** [comment]
After successfully installing the model-eos3b5e using the following command:
`ersilia -v fetch eos3b5e`
`ersilia serve eos3b5e` and 
`ersilia -v api run -i "CCCC"`

and got the following output:
{
"input": {
"key": "IJDNQMDRQITEOD-UHFFFAOYSA-N",
"input": "CCCC",
"text": "CCCC"
},
"output": {
"mw": 58.123999999999995
}
}

**Comment by Richiio** [comment]
### **My motivation**

Hello everyone, I'm Sarima Chiorlu. I am a machine learning engineer hoping to make contributions in medicine and drug discovery. Ersilia goal in making research for drugs easier for researchers in different regions was my major selling point to choosing to work with them as an intern. Using AI/ML, I can easily get information on things like protein binding, discover the elements present in a drug molecule and make research faster.

I come from Nigeria and apart from the debilitating state of our healthcare, we are often faced with disease challenges, whether it is global as in the COVID-19 or it is affects only our country like the ebola virus. Our government doesn't invest in the healthcare sector. However, through the work that you do, researchers can now easily through machine learning identify proteins, etc in drug substances.

To further understand why I am interested in this project, we see that Nigeria has a high number of Sickle cell individuals, where we see that what kills them are the silliest of disease. Dealing with a sickle cell relative is hardwork, this is because their white blood count is quite low and they are prone to the simplest of disease. But your work in not only investing in drug discovery but drug discovery in regions ensures that these people stand a higher chance.  I admire how Ersilia ensures that knowledge is not just transferred but also cultivated within these regions, fostering lasting impact. I wish to learn more about drug discovery while still working on your project and tasks. I firmly believe that the democratization of scientific tools and data-driven insights is necessary towards addressing some of the world's most pressing challenges.

I just wish to use my little knowledge to contribute to this vision and goal. I look forward to contributing to this project during my internship and also outside of the internship.

**Comment by Richiio** [comment]
First contribution just made

**Comment by Richiio** [comment]
I'm open to helping testing out some models where necessary.

**Comment by Richiio** [comment]
### **WEEK 2**

**Comment by DhanshreeA** [comment]
Hi @Richiio, thanks for the detailed updates. You can go ahead and start looking at Week 2 tasks.

**Comment by Richiio** [comment]
Hi @DhanshreeA, it's been updated. I was still working on putting it together. For week 2, I have decided to work on the Smiles-to-iUPAC translator which can be found [here](https://github.com/Kohulan/Smiles-TO-iUpac-Translator)

**Comment by DhanshreeA** [comment]
Thanks for the updates @Richiio

**Comment by Richiio** [comment]
### **Smiles to IUPAC Translator**

Brief Background
SMILES: which stands for Simplified Molecular Input Line Entry System. It is a widely used in the field of drug discovery and chemistry for representing chemical structures and molecules in a human readable form.
IUPAC: Our good old way of representing organic compounds (take a step back to organic chemistry in secondary school, most of us hated naming it, we know.). You take into account, the bonds present in these compounds. However, generally, it helps researchers categorize and analyze drugs within specific chemical classes, which can be important for understanding their properties and potential applications.

To begin, I am working with Google Colab, the implementation is straight-forward, and a lot of dependencies are handled for you.
I first ran the command: `!pip install STOUT-pypi`
It should be noted that after running this command, you have to re-start your runtime to ensure all old dependencies that were uninstalled are no longer there, this is because if you don't you would run into errors along the way
Next, to test my installation, I ran the command :
```
from STOUT import translate_forward, translate_reverse

# SMILES to IUPAC name translation

SMILES = "CN1C=NC2=C1C(=O)N(C(=O)N2C)C"
IUPAC_name = translate_forward(SMILES)
print("IUPAC name of "+SMILES+" is: "+IUPAC_name)

# IUPAC name to SMILES translation

IUPAC_name = "1,3,7-trimethylpurine-2,6-dione"
SMILES = translate_reverse(IUPAC_name)
print("SMILES of "+IUPAC_name+" is: "+SMILES)
```

And got the following Output:
```
Downloading trained model to /root/.data/STOUT-V2/models
/root/.data/STOUT-V2/models.zip
... done downloading trained model!
IUPAC name of CN1C=NC2=C1C(=O)N(C(=O)N2C)C is: 1,3,7-trimethylpurine-2,6-dione
SMILES of 1,3,7-trimethylpurine-2,6-dione is: CN1C=NC2=C1C(=O)N(C)C(=O)N2C
```
This showed that all was working as they should.

**Comment by Richiio** [comment]
Using the dataset provided for us in the internship handbook, the Essentials Medicine List which can be found [here](https://[raw.githubusercontent.com/ersilia-os/ersilia/master/notebooks/eml_canonical.csv](https://raw.githubusercontent.com/ersilia-os/ersilia/master/notebooks/eml_canonical.csv))

To get started, I began by first saving the dataset to my local machine. This can be achieved in two ways:
1. Clicking on the link, it opens then you proceed to right-clicking on the file and click on save As. Provide a name and Click Save (which is the easiest approach) or
2. Downloading it directly from ersilia repo, Navigate to ersilia-os, then to ersilia then to notebooks. You would see the dataset, once open, there is a download button on the top-right area. You click download and you have your dataset.

I began by first running predictions on the main model. I had wanted to create a new column to save the can_smiles_to_IUPAC and smiles_to_IUPAC values for all the entries in the dataset. However, due to how large the dataset was. it took a long time to provide results. A time(>6hrs). I stopped the process and proceeded to taking the first 20 rows in my data. This time, I saved the smiles_to_IUPAC values in a new dataset which I called smiles and the can_smiles_to_IUPAC into a new dataset called can_smiles. The code I used can be found here:

```
import pandas as pd
from STOUT import translate_forward

# Loading data from the CSV file
df = pd.read_csv('/content/eml_canonical.csv')

# Selecting the first 20 rows
df = df.head(20)

# Create empty DataFrames for SMILES-to-IUPAC and CAN-SMILES-to-IUPAC translations
smiles_IUPAC_name = pd.DataFrame(columns=['drugs', 'smiles', 'iupac_name'])
can_smiles_IUPAC_name = pd.DataFrame(columns=['drugs', 'can_smiles', 'iupac_name'])

# Translate SMILES to IUPAC and append to smiles_IUPAC_name
for index, row in df.iterrows():
    iupac_name = translate_forward(row['smiles'])
    smiles_IUPAC_name = smiles_IUPAC_name.append({'drugs': row['drugs'], 'smiles': row['smiles'], 'iupac_name': iupac_name}, ignore_index=True)

# Translate CAN-SMILES to IUPAC and append to can_smiles_IUPAC_name
for index, row in df.iterrows():
    iupac_name = translate_forward(row['can_smiles'])
    can_smiles_IUPAC_name = can_smiles_IUPAC_name.append({'drugs': row['drugs'], 'can_smiles': row['can_smiles'], 'iupac_name': iupac_name}, ignore_index=True)

# Save the DataFrames to Excel files
smiles_IUPAC_name.to_excel('smiles.xlsx', index=False)
can_smiles_IUPAC_name.to_excel('can_smiles.xlsx', index=False)
```
The output files generated can be found below:
[can_smiles.xlsx](https://github.com/ersilia-os/ersilia/files/12837465/can_smiles.xlsx)
[smiles.xlsx](https://github.com/ersilia-os/ersilia/files/12837466/smiles.xlsx)

**Comment by Richiio** [comment]
To compare the models with Ersilia's Hub implementation. I had to first identify the model name in Ersilia's repo. Initially I was trying to fetch model [eos5ecc](https://github.com/ersilia-os/eos5ecc). I began fetching the model but encountered some errors midway. The log can be found below:

[first_model_log.txt](https://github.com/ersilia-os/ersilia/files/12838184/first_model_log.txt)

After some time looking through it and reading the error message. I got to realise that I was getting the following error because the metadata.json file was empty. This file needed to have entries and not be empty. I contemplated raising an issue for this but decided to check if there was a second implementation of the model with it's metadata filled. I discovered the [eos4se9](https://github.com/ersilia-os/eos4se9) model which showed more activity history in comparison to the latter and had its metadata filled in. I tried fetching it and it completed without any issue.

@DhanshreeA would it be better to close model eos5ecc to prevent further confusions in the future. Just my suggestion though :)

**Comment by Richiio** [comment]
After I had successfully fetched the model and served it. I began running predictions on my Essential Medicines List dataset. Before I began, I created a folder within my ersilia I cloned which I named data, I copy-pasted the datasets I got from my earlier generated output. The `can_smiles` and `smiles` dataset. This dataset was initially saved in microsoft excel spreadsheet format, this had to be changed as Ersilia only supports comma separated(csv) files, json files, .tsv or .hdf5.

I ran this commands to generate the output:

`ersilia api run -i smiles.csv -o output.csv`
This produced the output below:
[output.csv](https://github.com/ersilia-os/ersilia/files/12838467/output.csv)  - Ersilia's prediction for smiles_to_IUPAC

`ersilia api run -i can_smiles.csv -o output2.csv`
This produced the output below:
[output2.csv](https://github.com/ersilia-os/ersilia/files/12838472/output2.csv)  - Ersilia's prediction for can_smiles_to_IUPAC

From the outputs generated from the original code and Ersilia's model implementation. Some are slightly close while some are the same. Let's have a look at the first 4 smiles from the Essential Medicine List to have a look at the research.

For the smiles, we have:
Drug: abacavir
Original code: `[(1S,4R)-4-[2-amino-6-(cyclopropylamino)purin-9-yl]cyclopent-2-en-1-yl]methanol`
Ersilia: 
```
{
    "input": {
        "key": "MCGSCOLBFJQGHM-SCZZXKLOSA-N",
        "input": "Nc1nc(NC2CC2)c3ncn([C@@H]4C[C@H](CO)C=C4)c3n1",
        "text": "Nc1nc(NC2CC2)c3ncn([C@@H]4C[C@H](CO)C=C4)c3n1"
    },
    "output": {
        "outcome": [
            "[(1R,4R)-4-[2-amino-4-(cyclopropylamino)-4H-purin-9-yl]cyclopent-2-en-1-yl]methanol"
        ]
    }
}
```

Drug: abiraterone
Original code: `(3S,8R,9S,10R,13S,14S)-10,13-dimethyl-17-pyridin-3-yl-2,3,4,7,8,9,11,12,14,15-decahydro-1H-cyclopenta[a]phenanthren-3-ol`
Ersilia:
```
{
    "input": {
        "key": "GZOSMCIZMLWJML-VJLLXTKPSA-N",
        "input": "C[C@]12CC[C@H](O)CC1=CC[C@@H]3[C@@H]2CC[C@@]4(C)[C@H]3CC=C4c5cccnc5",
        "text": "C[C@]12CC[C@H](O)CC1=CC[C@@H]3[C@@H]2CC[C@@]4(C)[C@H]3CC=C4c5cccnc5"
    },
    "output": {
        "outcome": [
            "(1S,2S,5S,10R,11R,14S)-5,11-dimethyl-5-pyridin-3-yltetracyclo[9.4.0.02,6.010,14]pentadeca-7,16-dien-14-ol"
        ]
    }
}
```

Drug: acetazolamide
Original code: ` N-(5-sulfamoyl-1,3,4-thiadiazol-2-yl)acetamide`
Ersilia:
```
{
    "input": {
        "key": "BZKPWHYZMXOIDC-UHFFFAOYSA-N",
        "input": "CC(=O)Nc1sc(nn1)[S](N)(=O)=O",
        "text": "CC(=O)Nc1sc(nn1)[S](N)(=O)=O"
    },
    "output": {
        "outcome": [
            "N-[5-[amino(dioxo)-\u03bb6-thia-3,4-diazacyclopent-2-en-2-yl]acetamide"
        ]
    }
}
```

Drug: acetic acid
Original code: `aceticacid`
Ersilia:
```
{
    "input": {
        "key": "QTBSBXVTEAMEQO-UHFFFAOYSA-N",
        "input": "CC(O)=O",
        "text": "CC(O)=O"
    },
    "output": {
        "outcome": [
            "aceticacid"
        ]
    }
}
```

**Comment by Richiio** [comment]
@DhanshreeA Are the results not meant to be the same consistently?

**Comment by Richiio** [comment]
### **Install and run docker**
This was my first time working with docker. So I began with the installation. 

```
# Updating package lists
sudo apt update
```
Next, I proceeded with installing the required packages to set up the Docker repository
```
sudo apt install -y apt-transport-https ca-certificates curl software-properties-common
```

Next, you add the Docker's official GPG key
```
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
```

Next, I added the Docker repository to APT sources
```
echo "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
```

Then we Install Docker Engine
```
sudo apt update
sudo apt install -y docker-ce docker-ce-cli containerd.io
```

Start the Docker service
```
sudo service docker start
```

**Comment by Richiio** [comment]
To confirm my docker installation, I ran the following command `docker ps` and got the result shown below:

```
(base) root@Richio:~# docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
```
Next, I decided to try pulling a model from Ersilia's 
```
(base) root@Richio:~# docker pull ersiliaos/eos4se9
Using default tag: latest
latest: Pulling from ersiliaos/eos4se9
8b91b88d5577: Pull complete
824416e23423: Pull complete
bbe2c2981082: Pull complete
7b6b68d15a5c: Pull complete
71f8f4db541d: Pull complete
4f4fb700ef54: Pull complete
278266b40c52: Pull complete
4298588a86ad: Pull complete
dddca77c0f59: Pull complete
a113a2030c72: Pull complete
0c8571d61669: Pull complete
Digest: sha256:3c0b4dab7a313bfb33c74b45ca378f7d69b0b9dbaaf843357780180910af31ab
Status: Downloaded newer image for ersiliaos/eos4se9:latest
docker.io/ersiliaos/eos4se9:latest
```

**Comment by Richiio** [comment]
### Smiles2IUPAC translator model Explanation

### **Understanding Ersilia's Backend**

**Comment by Richiio** [comment]
## **ImageMol** Installation guide

@DhanshreeA mentioned we can use CPU instead of GPU or CUDA

Creating our ImageMol environment and activating it
Step 1: ` conda create -n imagemol python=3.7.3`\
Step 2:  `conda activate imagemol`
Step 3: `conda install -c rdkit rdkit`

Install Torch, Torchvision. We picked the package compatible with the cpu
Step 4: `pip install https://download.pytorch.org/whl/cpu/torch-1.4.0%2Bcpu-cp37-cp37m-linux_x86_64.whl`
Step 5:  `pip install https://download.pytorch.org/whl/cpu/torchvision-0.5.0%2Bcpu-cp37-cp37m-linux_x86_64.whl`
Step 6:  `pip install torch torchvision torchaudio`

Installing the required packages
Step 7: `pip install torch-scatter`
Step 8: `pip install torch-cluster`
Step 9: `pip install torch-sparse`
Step 10: `pip install torch-spline-conv -f https://pytorch-geometric.com/whl/torch-1.4.0.html`
Step 11: `mkdir ckpts`
Step 12: `wget -P ckpts/ https://github.com/HongxinXiang/ImageMol/blob/master/ckpts/pretraining-toy/checkpoints/ImageMol_10.pth.tar
`
Step 13: `sudo apt update`
Step 14: `sudo apt install libxrender1`
Step 15: `git clone https://github.com/HongxinXiang/ImageMol.git`
Step 16: cd ImageMol, run `pip install -r  requirements.txt`, cd datasets, cd finetuning
Step 17: Pick a dataset you want to finetune on and download it into the finetuning folder, we would be working with the SARS-CoV-2. 
Step 18: `pip install gdown`
Step 19: `gdown --id 1UfROoqR_aU6f5xWwxpLoiJnkwyUzDsui #(id of the google file, this is basically the value between the d/ and before the /view`
Step 20: `mkdir SARS-CoV-2`
Step 21: `tar -xzvf SARS-CoV-2.tar.gz -C SARS-CoV-2`

Finetuning our dataset on the pretrained model
Step 22:
 ```
python finetune.py --gpu 0 \
                   --save_finetune_ckpt 1 \
                   --log_dir ./logs/toxcast \
                   --dataroot ./datasets/finetuning/SARS-CoV-2/SARS-CoV-2 \
                   --dataset 3CL_enzymatic_activity\
                   --task_type classification \
                   --resume ./ckpts/ImageMol.pth.tar \
                   --image_aug \
                   --lr 0.5 \
                   --batch 64 \
                   --epochs 20
```

**Comment by DhanshreeA** [comment]
> To compare the models with Ersilia's Hub implementation. I had to first identify the model name in Ersilia's repo. Initially I was trying to fetch model [eos5ecc](https://github.com/ersilia-os/eos5ecc). I began fetching the model but encountered some errors midway. The log can be found below:
> 
> [first_model_log.txt](https://github.com/ersilia-os/ersilia/files/12838184/first_model_log.txt)
> 
> After some time looking through it and reading the error message. I got to realise that I was getting the following error because the metadata.json file was empty. This file needed to have entries and not be empty. I contemplated raising an issue for this but decided to check if there was a second implementation of the model with it's metadata filled. I discovered the [eos4se9](https://github.com/ersilia-os/eos4se9) model which showed more activity history in comparison to the latter and had its metadata filled in. I tried fetching it and it completed without any issue.
> 
> @DhanshreeA would it be better to close model eos5ecc to prevent further confusions in the future. Just my suggestion though :)

Hi @Richiio thank you for your suggestion. I should point out that the two models are different in that they do the opposite of each other. `eos4se9` takes SMILES inputs and generates IUPAC names (tex outputs), whereas `eos5ecc` takes in IUPAC names (text inputs) and generates SMILES. Incorporating text inputs within Ersilia is still very much WIP hence this model is in the backlog to be implemented.

**Comment by DhanshreeA** [comment]
> @DhanshreeA Are the results not meant to be the same consistently?

@Richiio could you report which version of STOUT you are using? Ersilia is currently using `2.0.1`. However the latest version of STOUT seems to be 2.0.6. If the base model within this package has been updated (whether in terms of its architecture or the training dataset), it is possible that there are differences across the two versions. We will have to see which is more accurate.

**Comment by DhanshreeA** [comment]
@Richiio I also noticed that you worked with STOUT on Google Colab, could you mention why? Were there issues running it on your machine?

**Comment by Richiio** [comment]
Thanks @DhanshreeA for the updates! I did encounter two challenges when installing the model. However, the challenges encountered occurred when I was working outside the conda environment. 
All worked fine within the conda environment apart from the `OSError: [Errno 0] JVM DLL not found: Define/path/or/set/JAVA_HOME/variable/properly` error which you would encounter if you do not have Java installed and added to your environment path and if you decide to use the `conda install -c decimer stout-pypi`. I created the STOUT environment and installed STOUT-pypi using the `pip` command instead of `conda`

The JVM error, I encountered was resolved by installing java using the command `sudo apt install openjdk-8-jre`  and setting JAVA to the environment variable which was done using the following commands: 
`nano ~/.bashrc` which opened the shell profile file in text editor and added the following command at the end of the file `export JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64` . I saved the file and applied the changes using this command `source ~/.bashrc` 

I created a file which I named sarima.py inside a folder I created, downloaded the dataset to the same file location as the python file and ran `nano sarima.py`, this opened the shell and I pasted the following code:
```
import pandas as pd
from STOUT import translate_forward

# Loading data from the CSV file (using a relative path)
csv_file_path = 'eml_canonical.csv'
df = pd.read_csv(csv_file_path)

# Selecting the first 20 rows
df = df.head(20)

# Create empty DataFrames for SMILES-to-IUPAC and CAN-SMILES-to-IUPAC translations
smiles_IUPAC_name = pd.DataFrame(columns=['drugs', 'smiles', 'iupac_name'])
can_smiles_IUPAC_name = pd.DataFrame(columns=['drugs', 'can_smiles', 'iupac_name'])

# Translate SMILES to IUPAC and concatenate with smiles_IUPAC_name
for index, row in df.iterrows():
    iupac_name = translate_forward(row['smiles'])
    smiles_IUPAC_name = pd.concat([smiles_IUPAC_name, pd.DataFrame({'drugs': [row['drugs']], 'smiles': [row['smiles']], 'iupac_name': [iupac_name]})], ignore_index=True)

# Translate CAN-SMILES to IUPAC and concatenate with can_smiles_IUPAC_name
for index, row in df.iterrows():
    iupac_name = translate_forward(row['can_smiles'])
    can_smiles_IUPAC_name = pd.concat([can_smiles_IUPAC_name, pd.DataFrame({'drugs': [row['drugs'], 'can_smiles': [row['can_smiles']], 'iupac_name': [iupac_name]})], ignore_index=True)

# Save the DataFrames to CSV files
smiles_IUPAC_name.to_csv('smiles.csv', index=False)
can_smiles_IUPAC_name.to_csv('can_smiles.csv', index=False)
```
saved it and exited from the shell. I installed pandas and ran `python sarima.py`. This ran and produced the following output
[can_smiles.csv](https://github.com/ersilia-os/ersilia/files/12873172/can_smiles.csv)
[smiles.csv](https://github.com/ersilia-os/ersilia/files/12873173/smiles.csv)

I decided to work with Colab as it as a lot easier and faster
  

> @Richiio I also noticed that you are working with models on Google Colab, could you mention why? Were there issues running them on your machine?

**Comment by Richiio** [comment]
Even using the above installation, the output was different to that of Ersilia's  Model Hub

Model Version
```
(STOUT) root@Richio:~/Test#  pip show stout-pypi
Name: STOUT-pypi
Version: 2.0.5
Summary: STOUT V2.0 - Smiles TO iUpac Translator Version 2.0
Home-page: https://github.com/Kohulan/Smiles-TO-iUpac-Translator
Author: Kohulan Rajan
Author-email: kohulan.rajan@uni-jena.de
License: MIT
Location: /root/miniconda3/envs/STOUT/lib/python3.8/site-packages
Requires: jpype1, pystow, tensorflow, unicodedata2
Required-by:
```

Ersilia's version:
```
(ersilia) root@Richio:~/Test# conda list
# packages in environment at /root/miniconda3/envs/ersilia:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main
_openmp_mutex             5.1                       1_gnu
alembic                   1.12.0                   pypi_0    pypi
attrs                     21.4.0                   pypi_0    pypi
bentoml                   0.11.0                   pypi_0    pypi
blinker                   1.6.2                    pypi_0    pypi
boto3                     1.28.61                  pypi_0    pypi
botocore                  1.31.61                  pypi_0    pypi
bzip2                     1.0.8                h7b6447c_0
ca-certificates           2023.7.22            hbcca054_0    conda-forge
cerberus                  1.3.5                    pypi_0    pypi
certifi                   2023.7.22                pypi_0    pypi
chardet                   5.2.0                    pypi_0    pypi
charset-normalizer        3.3.0                    pypi_0    pypi
chembl-webresource-client 0.10.8                   pypi_0    pypi
click                     8.1.7                    pypi_0    pypi
docker                    6.1.3                    pypi_0    pypi
dockerfile-parse          2.0.1                    pypi_0    pypi
easydict                  1.10                     pypi_0    pypi
emoji                     2.8.0                    pypi_0    pypi
ersilia                   0.1.27                   pypi_0    pypi
```

**Comment by Richiio** [comment]
> > To compare the models with Ersilia's Hub implementation. I had to first identify the model name in Ersilia's repo. Initially I was trying to fetch model [eos5ecc](https://github.com/ersilia-os/eos5ecc). I began fetching the model but encountered some errors midway. The log can be found below:
> > [first_model_log.txt](https://github.com/ersilia-os/ersilia/files/12838184/first_model_log.txt)
> > After some time looking through it and reading the error message. I got to realise that I was getting the following error because the metadata.json file was empty. This file needed to have entries and not be empty. I contemplated raising an issue for this but decided to check if there was a second implementation of the model with it's metadata filled. I discovered the [eos4se9](https://github.com/ersilia-os/eos4se9) model which showed more activity history in comparison to the latter and had its metadata filled in. I tried fetching it and it completed without any issue.
> > @DhanshreeA would it be better to close model eos5ecc to prevent further confusions in the future. Just my suggestion though :)
> 
> Hi @Richiio thank you for your suggestion. I should point out that the two models are different in that they do the opposite of each other. `eos4se9` takes SMILES inputs and generates IUPAC names (text outputs), whereas `eos5ecc` takes in IUPAC names (text inputs) and generates SMILES. Incorporating text inputs within Ersilia is still very much WIP hence this model is in the backlog to be implemented.

Thanks for the clarification!

**Comment by Richiio** [comment]
I created a new STOUT environment which I called STOUT2, activated it and ran the command pip install STOUT-pypi==2.0.1, downloaded the dataset and ran my python script. 
The number of SMILES to IUPAC predicted correctly improved but we still had some incorrect predictions. Those incorrect predictions, to confirm which was right and wrong. I went to pubchem and searched for the drug using its smile which showed the correct IUPAC name as seen [here ](https://pubchem.ncbi.nlm.nih.gov/compound/441300) . The result corresponded with that of the original model.
<html>
<body>
<google-sheets-html-origin><style type="text/css"></style>

drugs | smiles | iupac_name
-- | -- | --
abacavir | Nc1nc(NC2CC2)c3ncn([C@@H]4C[C@H](CO)C=C4)c3n1 | [(1S,4R)-4-[6-(cyclopropylamino)-2-(methylamino)-6H-purin-9-yl]cyclopent-2-en-1-yl]methanol
abiraterone | C[C@]12CC[C@H](O)CC1=CC[C@@H]3[C@@H]2CC[C@@]4(C)[C@H]3CC=C4c5cccnc5 | (3S,6aS,6bR,10aR,10bS)-6a,10a-dimethyl-7-pyridin-3-yl-1,2,3,4,6,7,10,10b-octahydrobenzo[a]azulen-3-ol
acetazolamide | CC(=O)Nc1sc(nn1)[S](N)(=O)=O | N-[5-[amino(dioxo)-λ6-sulfanyl]-1,3,4-thiadiazol-2-yl]acetamide
acetic acid | CC(O)=O | aceticacid
acetylcysteine | CC(=O)N[C@@H](CS)C(O)=O | (2R)-2-acetamido-3-sulfanylpropanoicacid
acetylsalicylic acid | CC(=O)Oc1ccccc1C(O)=O | 2-acetyloxybenzoicacid
aciclovir | NC1=NC(=O)c2ncn(COCCO)c2N1 | 9-(2-hydroxyethoxymethyl)-2-(methylamino)-3H-purin-6-one
aclidinium | OC(C(=O)O[C@H]1C[N+]2(CCCOC3=CC=CC=C3)CCC1CC2)(C1=CC=CS1)C1=CC=CS1 | 2-[[(3R)-1-(3-phenoxypropyl)-1-azoniabicyclo[2.2.2]octan-3-yl]oxy]-1,1-dithiophen-2-ylethanol
amlodipine | CCOC(=O)C1=C(COCCN)NC(=C(C1c2ccccc2Cl)C(=O)OC)C | 3-O-ethyl5-O-methyl2-(2-aminoethoxymethyl)-4-(6-chlorophenyl)-6-methyl-1,4-dihydropyridine-3,5-dicarboxylate
amodiaquine | CCN(CC)Cc1cc(Nc2ccnc3cc(Cl)ccc23)ccc1O | 4-[(7-chloroquinolin-4-yl)amino]-2-(diethylaminomethyl)cyclohexa-1,3,5-trien-1-ol


</body>
</html>

**Comment by Richiio** [comment]
# **WEEK 3**
## Model Suggestions

### Model 1
DiffDock: Diffusion Steps, Twists, and Turns for Molecular Docking

**Model Description** 
DiffDock is a model for molecular docking in drug discovery. It takes a fresh approach by treating docking as a generative modeling problem, which helps it achieve better accuracy. DiffDock efficiently explores the possibilities of how small molecules bind to proteins by using a unique diffusion process. DiffDock outperforms traditional docking methods and deep learning techniques, achieving a 38% success rate on the PDBBind dataset, even excelling in computationally folded structures.

Model Identifier
Slug: DiffDock

Model Characteristics
Input: Ligand poses
Task: Molecular docking
Tag: Drug discovery, Protein-ligand interaction
Output: Binding structure, Score

References
[Source Code](https://github.com/gcorso/DiffDock)
[Publication](https://arxiv.org/abs/2210.01776)


License: [MIT License](https://github.com/gcorso/DiffDock/blob/main/LICENSE)

**Comment by Richiio** [comment]
### **Model 2**

**SELFormer: Molecular Representation Learning via SELFIES Language Models**

**Comment by Richiio** [comment]
Model 3

Deep Docking: A Deep Learning Platform for Augmentation of Structure Based Drug Discovery

Model Description
Drug discovery is a costly and time-intensive endeavor, hindered by the rapid growth of chemical databases. To tackle this challenge, Deep Docking (DD) is introduced as a deep learning platform capable of rapidly and accurately docking billions of molecular structures. DD employs quantitative structure-activity relationship (QSAR) models trained on subsets of a chemical library to predict docking outcomes, efficiently removing unfavorable molecules iteratively. When combined with the FRED docking program, DD calculates docking scores for 1.36 billion molecules, achieving remarkable data reduction and enrichment of high-scoring molecules without sacrificing favorable results. DD's flexibility allows integration with any docking program, and its resources are publicly available.

Model Identifier
Slug: DeepDock

Model Characteristics
Input: SMILES
Task: Molecular docking
Tag: Drug discovery
Output: Docked SMILES

References
[Source Code](https://github.com/vibudh2209/D2)
[Publication](https://pubs.acs.org/doi/10.1021/acscentsci.0c00229#)

License: [MIT License](https://github.com/vibudh2209/D2/blob/master/LICENSE.txt)

**Comment by Richiio** [comment]
Model

DeepAffinity: interpretable deep learning of compound–protein affinity through unified recurrent and convolutional neural networks

Model description

Model Identifier
Slug: DeepAffinity

Model Characteristics
Input: 
Task:
Tag: 
Output: 

References
[Source Code](https://github.com/Shen-Lab/DeepAffinity)
[Publication](https://academic.oup.com/bioinformatics/article/35/18/3329/5320555)

License: [GPL-3.0 License](https://github.com/Shen-Lab/DeepAffinity/blob/master/LICENSE)

**Comment by DhanshreeA** [comment]
@Richiio Additionally, thank you for the model suggestions, you can mark week 3 tasks as completed. :)

**Comment by Richiio** [comment]
Thanks so much @DhanshreeA for getting back. For the extra task, I created a third environment which I called STOUT3 and installed STOUT-pypi using the `pip install git+https://github.com/Kohulan/Smiles-TO-iUpac-Translator.git`. To confirm I was using the right version. I ran the pip show command and got the following output

```
(STOUT3) root@Richio:~# pip show stout-pypi
Name: STOUT-pypi
Version: 2.0.6
Summary: STOUT V2.0 - Smiles TO iUpac Translator Version 2.0
Home-page: https://github.com/Kohulan/Smiles-TO-iUpac-Translator
Author: Kohulan Rajan
Author-email: kohulan.rajan@uni-jena.de
License: MIT
Location: /root/miniconda3/envs/STOUT3/lib/python3.8/site-packages
Requires: jpype1, pystow, tensorflow, unicodedata2
Required-by:
```

I created the columns for the various versions, retrieved the csv file, then proceeded to excel to create columns for the PubChem reference, I copy-pasted the SMILES and got their various IUPAC names. The final csv compilation can be found here:

[Final Result - smiles (1).csv.csv](https://github.com/ersilia-os/ersilia/files/13035166/Final.Result.-.smiles.1.csv.csv)

The version incorrect with that of pubchem was 2.0.1, versions 2.0.5 and 2.0.6 corresponded with the results from Pubchem(original code)
